File: 09/day9.pl
:- use_module(library(clpfd)).
:- dynamic mem/2.

run :-
    parse(Program),
    run_to_halt(Program, [1], Ans1),
    format('Part 1: ~w~n', Ans1),
    run_to_halt(Program, [2], Ans2),
    format('Part 2: ~w~n', Ans2).

parse(Input) :-
    csv_read_file("day9.input", [Row], []),
    Row =.. [_|Input].

enumerate([], [], _).
enumerate([H|T], [I-H|TT], I) :-
    NI #= I + 1,
    enumerate(T, TT, NI).

list_to_mem(List) :-
    enumerate(List, OrdPairs, 0),
    maplist([I-X]>>assertz(mem(I, X)), OrdPairs).

set(Index, Value) :-
    retractall(mem(Index,_)),
    assertz(mem(Index, Value)).

get(Index, Value) :-
    ( mem(Index, X) 
    ->
        Value #= X
    ;
        Value #= 0
    ).

mode(0, Index, Pos, _, Value) :-
    mode(1, Index, Pos, _, X),
    get(X, Value).

mode(1, Index, Pos, _, Value) :-
    I #= Index + Pos,
    get(I, Value).

mode(2, Index, Pos, RelativeBase, Value) :-
    mode(1, Index, Pos, _, X),
    Rel #= X + RelativeBase,
    get(Rel, Value).

value(First, Index, Pos, RelBase, Value) :-
    Mod #= (First // (10 * (10^Pos))) mod 10,
    Mod in 0..2,
    mode(Mod, Index, Pos, RelBase, Value).

addr(First, Index, Pos, RelBase, Value) :-
    Mod #= (First // (10 * (10^Pos))) mod 10,
    I #= Index + Pos,
    ( Mod #= 0 
    ->
        get(I, Value)
    ;
        get(I, X),
        Value #= X + RelBase
    ).

value_value(First, Index, RelBase, X, Y) :-
    value(First, Index, 1, RelBase, X),
    value(First, Index, 2, RelBase, Y).

value_value_addr(First, Index, RelBase, X, Y, Z) :-
    value(First, Index, 1, RelBase, X),
    value(First, Index, 2, RelBase, Y),
    addr(First, Index, 3, RelBase, Z).

opcode(1, First, I, I, Index, NewIndex, RB, RB, _, false) :-
    value_value_addr(First, Index, RB, X, Y, Z),
    New #= X+Y, 
    set(Z, New), 
    NewIndex #= Index+4. 

opcode(2, First, I, I, Index, NewIndex, RB, RB, _, false) :-
    value_value_addr(First, Index, RB, X, Y, Z),
    New #= X*Y, 
    set(Z, New), 
    NewIndex #= Index+4. 

opcode(3, First, [Input|RemInput], RemInput, Index, NewIndex, RB, RB, _, false) :-
    addr(First, Index, 1, RB, X),
    set(X, Input), 
    NewIndex #= Index+2. 

opcode(4, First, I, I, Index, NewIndex, RB, RB, Output, false) :-
    value(First, Index, 1, RB, Output),
    NewIndex #= Index+2. 

opcode(5, First, I, I, Index, NewIndex, RB, RB, _, false) :-
    value_value(First, Index, RB, X, Y),
    ( X #\= 0
    ->
        NewIndex #= Y
    ;
        NewIndex #= Index+3
    ).

opcode(6, First, I, I, Index, NewIndex, RB, RB, _, false) :-
    value_value(First, Index, RB, X, Y),
    ( X #= 0
    ->
        NewIndex #= Y
    ;
        NewIndex #= Index+3
    ).

opcode(7, First, I, I, Index, NewIndex, RB, RB, _, false) :-
    value_value_addr(First, Index, RB, X, Y, Z),
    ( X #< Y
    ->
        set(Z, 1)
    ;
        set(Z, 0)
    ),
    NewIndex #= Index+4. 

opcode(8, First, I, I, Index, NewIndex, RB, RB, _, false) :-
    value_value_addr(First, Index, RB, X, Y, Z),
    ( X #= Y
    ->
        set(Z, 1)
    ;
        set(Z, 0)
    ),
    NewIndex #= Index+4. 

opcode(9, First, I, I, Index, NewIndex, RelBase, NewRelBase, _, false) :-
    value(First, Index, 1, RelBase, X),
    NewRelBase #= RelBase + X,
    NewIndex #= Index+2.

opcode(99, _, _, _, _, _, _, _, 0, true).

run(Input, RemInput, Index, NewIndex, RelBase, NewRelBase, Output, ProgramHalted) :-
    get(Index, First),
    Opcode #= First mod 100,
    opcode(Opcode, First, Input, NInput, Index, NIndex, RelBase, NRelBase, Out, Halted),
    ( not(var(Out))
    ->
        ( Halted=true
        ->
            ProgramHalted=true
        ;
            ProgramHalted=false,
            Output=Out 
        ),
        RemInput=NInput,
        NewIndex=NIndex,
        NewRelBase=NRelBase
    ;
        run(NInput, RemInput, NIndex, NewIndex, NRelBase, NewRelBase, Output, ProgramHalted)
    ).

run(Input, Output) :-
    run(Input, _, 0, _, 0, _, Output, _).

run_to_halt(Program, Input, Output) :-
    list_to_mem(Program),
    run_to_halt(Input, 0, 0, Output),
    retractall(mem(_,_)).

run_to_halt(Input, Index, RelBase, Output) :-
    run(Input, NewInput, Index, NewIndex, RelBase, NewRelBase, Out, Halted),
    ( Halted = true
    ->
        Output = []
    ;
        run_to_halt(NewInput, NewIndex, NewRelBase, NOut),
        Output = [Out|NOut]
    ).

:- begin_tests(part1).

test(quine) :-
    Program = [109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99],
    run_to_halt(Program, [], Output),
    assertion(Output = Program).

test(large_multiplication) :-
    Program = [1102,34915192,34915192,7,4,7,99,0],
    run_to_halt(Program, [], Ans),
    assertion(Ans == [1219070632396864]).

test(large_number) :-
    Program = [104,1125899906842624,99],
    run_to_halt(Program, [], Ans),
    assertion(Ans == [1125899906842624]).

:- end_tests(part1).

File: 09/day9.rs
use std::fs::File;
use std::io;
use std::io::Read;
use std::str::FromStr; use std::collections::VecDeque;
use std::collections::HashMap;

struct Program {
    mem: HashMap<i64, i64>,
    input: VecDeque<i64>,
    output: VecDeque<i64>,
    iptr: i64,
    relative_base: i64,
    halted: bool,
    error: bool
}

impl Program {
    fn new(program: &HashMap<i64, i64>) -> Program {
        Program{
            mem: program.clone(),
            input: VecDeque::new(),
            output: VecDeque::new(),
            iptr: 0,
            relative_base: 0,
            halted: false,
            error: false
        }
    }

    fn give_input(&mut self, i: i64) {
        self.input.push_back(i);
    }

    fn get_output(&mut self) -> Option<i64> {
        self.output.pop_front()
    }

    fn run(&mut self) {
        self.run_program();
        if self.error {
            panic!();
        }
    }

    fn run_to_halt(&mut self) {
        while !self.halted {
            self.run();
        }
    }

    fn value(&mut self, raw: i64, pos: u32) -> i64 {
        let mode = (raw / (10_i64.pow(pos-1))) % 10;
        match mode {
            0 => self.value_direct(self.value_direct(self.iptr+pos as i64)),
            1 => self.value_direct(self.iptr+pos as i64),
            2 => self.value_direct(self.value_direct(self.iptr+pos as i64)+self.relative_base),
            _ => panic!()
        }
    }

    fn value_direct(&self, pos: i64) -> i64 {
        if pos < 0 {
            panic!()
        }
        *(self.mem.get(&pos).or(Some(&0)).unwrap())
    }

    fn addr(&mut self, raw: i64, pos: u32) -> i64 {
        let mode = (raw / (10_i64.pow(pos-1))) % 10;
        match mode {
            0 => self.value_direct(self.iptr+pos as i64),
            2 => self.value_direct(self.iptr+pos as i64)+self.relative_base,
            _ => panic!()
        }
    }

    fn run_program(&mut self) {
        loop {
            let opcode_raw = self.value_direct(self.iptr);
            let opcode = opcode_raw % 100;
            let modes = opcode_raw / 100;
            match opcode {
                1 => self.add(modes), 
                2 => self.multiply(modes),
                3 => self.read(modes),
                4 => { self.write(modes); return },
                5 => self.not_zero(modes), 
                6 => self.is_zero(modes), 
                7 => self.jump_if_less(modes), 
                8 => self.jump_if_equal(modes), 
                9 => self.set_relative_base(modes),
                99 => { self.halted = true; return },
                _ => { self.error = true; return },
            }
        }
    }

    fn add(&mut self, opcode_raw: i64) {
        let x = self.value(opcode_raw, 1);
        let y = self.value(opcode_raw, 2);
        let z = self.addr(opcode_raw, 3); 
        self.mem.insert(z, x+y);
        self.iptr += 4;
    }

    fn multiply(&mut self, opcode_raw: i64) {
        let x = self.value(opcode_raw, 1);
        let y = self.value(opcode_raw, 2);
        let z = self.addr(opcode_raw, 3); 
        self.mem.insert(z, x*y);
        self.iptr += 4;
    }

    fn read(&mut self, opcode_raw: i64) {
        let x = self.addr(opcode_raw, 1); 
        match self.input.pop_front() {
            Some(i) => self.mem.insert(x, i),
            None => panic!()
        };
        self.iptr += 2;
    }

    fn write(&mut self, opcode_raw: i64) {
        let x = self.value(opcode_raw, 1);
        self.output.push_back(x);
        self.iptr += 2;
    }

    fn not_zero(&mut self, opcode_raw: i64) {
        let x = self.value(opcode_raw, 1);
        let y = self.value(opcode_raw, 2);
        self.iptr = if x != 0 { y } else { self.iptr+3 };
    }

    fn is_zero(&mut self, opcode_raw: i64) {
        let x = self.value(opcode_raw, 1);
        let y = self.value(opcode_raw, 2);
        self.iptr = if x == 0 { y } else { self.iptr+3 };
    }

    fn jump_if_less(&mut self, opcode_raw: i64) {
        let x = self.value(opcode_raw, 1);
        let y = self.value(opcode_raw, 2);
        let z = self.addr(opcode_raw, 3); 
        if x < y { self.mem.insert(z, 1) } else { self.mem.insert(z, 0) };
        self.iptr += 4;
    }

    fn jump_if_equal(&mut self, opcode_raw: i64) {
        let x = self.value(opcode_raw, 1);
        let y = self.value(opcode_raw, 2);
        let z = self.addr(opcode_raw, 3); 
        if x == y { self.mem.insert(z, 1) } else { self.mem.insert(z, 0) };
        self.iptr += 4;
    }

    fn set_relative_base(&mut self, opcode_raw: i64) {
        self.relative_base += self.value(opcode_raw, 1);
        self.iptr += 2;
    }
}

fn parse(filename: &str) -> io::Result<HashMap<i64,i64>> {
    let mut file = File::open(filename)?;
    let mut input = String::new();
    file.read_to_string(&mut input)?;
    Ok(str_to_map(&input))
}

fn str_to_map(input: &str) -> HashMap<i64,i64> {
    let mut m = HashMap::new();
    for (i, x) in input.trim_end().split(',').map(|x| i64::from_str(&x).unwrap()).enumerate() {
        m.insert(i as i64, x);
    }
    m
}

fn main() {
    let input = parse("day9.input").unwrap();
    let mut program = Program::new(&input);
    program.give_input(1);
    program.run_to_halt();
    println!("Part 1: {}", program.get_output().unwrap());
    program = Program::new(&input);
    program.give_input(2);
    program.run_to_halt();
    println!("Part 2: {}", program.get_output().unwrap());
}

#[test]
fn test_quine_one_instr_per() {
    let p = str_to_map("109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99");
    let mut program = Program::new(p);
    program.set_relative_base(1);
    assert_eq!(program.relative_base, 1);
    program.write(2);
    assert_eq!(program.get_output(), Some(109));
}

#[test]
fn test_quine() {
    let p = str_to_map("109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99");
    let mut program = Program::new(p);
    program.run_to_halt();
    assert_eq!(program.output, vec![109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99]);
}

#[test]
fn test_16_digit() {
    let p = str_to_map("1102,34915192,34915192,7,4,7,99,0");
    let mut program = Program::new(p);
    program.run();
    assert_eq!(program.get_output(), Some(1219070632396864));
}

#[test]
fn test_large_number() {
    let p = str_to_map("104,1125899906842624,99");
    let mut program = Program::new(p);
    program.run();
    assert_eq!(program.get_output(), Some(1125899906842624));
}
